### PR TITLE
Runtime SPH_GRAD_RHO

### PR DESCRIPTION
**Pull request by sbird**
This removes the compile flag SPH_GRAD_RHO which is the LAST COMPILE TIME SCIENCE FLAG!! Hooray! No longer will we need to recompile MP-Gadget for different tests! Everything is runtime.

If you agree with the approach for sph_extra_data I might move some other things that are recomputed every timestep into it. eg, VelPred.

### CODE REVIEW

File: libgadget/run.c [hunk 1]
```diff
@@ -209,6 +211,8 @@ void run(DomainDecomp * ddecomp)
             cooling_and_starformation(&Tree);
         }
 
+        slots_free_sph_extra_data(SphP_extra);
```
Comment (rainwoodman) [inline libgadget/run.c]: What about calling this sph_scratch_data? 
It is some data that we do not need to preserve them; the word extra doesn't convey this message. However I do have the impression we probably use 'extra' for this meaning everywhere else anyways.

File: libgadget/sfr_eff.c [hunk 2]
```diff
@@ -830,7 +830,14 @@ find_star_mass(int i)
  * You may need a license to run with these modess.
 
  * */
-#if defined SPH_GRAD_RHO
+
+int need_to_compute_sph_grad_rho(void)
+{
+    if (HAS(sfr_params.StarformationCriterion, SFR_CRITERION_MOLECULAR_H2)) {
```
Comment (rainwoodman) [inline libgadget/sfr_eff.c]: Better to add an sfr prefix now than later, since it is not static.

File: libgadget/slotsmanager.h [hunk 3]
```diff
@@ -124,6 +128,9 @@ struct sph_particle_data
 #define BHP(i) BhP[P[i].PI]
 #define STARP(i) StarP[P[i].PI]
 
+/*Shortcut for the extra data*/
+#define SphP_extra ((struct sph_extra_data*) SlotsManager->info[0].extradata)
```
Comment (rainwoodman) [inline libgadget/slotsmanager.h]: Can we call it something like SphPE?

Comment (sbird) [reply libgadget/slotsmanager.h]: SphP_scratch. I don't like names that differ from another name by just one character.

**Comment by sbird** [comment]
Names done! Will test.

File: libgadget/slotsmanager.h (hunk 3 above)
Comment (rainwoodman) [reply libgadget/slotsmanager.h]: I guess I've been parsing too much pure consonant names...

Comment (sbird) [reply libgadget/slotsmanager.h]: Ah yes, the vowel makes a difference...I've seen several people confuse DESI and DES, for example.

**Comment by sbird** [comment]
Ok, a few segfaults later, this passes my tests...will merge when travis goes green.

**Pull request merged**